File: aroundHubSpringBoot/src/main/java/com/example/aroungHubJavaSpringBoot/aroundHubSpringBoot/controller/ProductController.java
package com.example.aroungHubJavaSpringBoot.aroundHubSpringBoot.controller;

import com.example.aroungHubJavaSpringBoot.aroundHubSpringBoot.data.dto.ProductDto;
import com.example.aroungHubJavaSpringBoot.aroundHubSpringBoot.service.ProductService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/product-api")
public class ProductController {
    private final Logger LOGGER = LoggerFactory.getLogger(ProductController.class);
    private ProductService productService;

    /*
    * Autowired
    *   -> 자동으로 의존성 연결해준다.
    * */
    @Autowired
    public ProductController(ProductService productService){
        this.productService = productService;
    }

    // http://localhost:8080/api/v1/product-api/product/{productId}
    @GetMapping(value = "/product/{productId}")
    public ProductDto getProduct(@PathVariable String productId){

        long startTime = System.currentTimeMillis();
        LOGGER.info("[ProductController] perform {} of Aroung Hub API.", "getProduct");

        ProductDto productDto = productService.getProduct(productId);

        LOGGER.info("[ProductController] Response :: productId = {}, productName = {}, productPrice = {}, productStock = {}, Response Time = {}ms",
                productDto.getProductId(), productDto.getProductName(), productDto.getProductPrice(), productDto.getProductStock(), (System.currentTimeMillis() - startTime));

        return productDto; 
    }

    // http://localhost:8080/api/v1/product-api/product
    @PostMapping("/product")
    public ProductDto createProduct(@RequestBody ProductDto productDto){

        String productId = productDto.getProductId();
        String productName = productDto.getProductName();
        int productPrice = productDto.getProductPrice();
        int productStock = productDto.getProductStock();

        return productService.saveProduct(productId, productName, productPrice, productStock);
    }

    // http://localhost:8080/api/v1/product-api/product/{productId}
    @DeleteMapping("/product/{productId}")
    public ProductDto deleteProduct(@PathVariable String productId){
        return null;
    }

}

File: aroundHubKotlinSpringBoot/src/main/kotlin/com/example/aroungHubKotlinSpringBoot/aroundHubKotlinSpringBoot/controller/ProductController.kt
package com.example.aroungHubKotlinSpringBoot.aroundHubKotlinSpringBoot.controller

import com.example.aroungHubKotlinSpringBoot.aroundHubKotlinSpringBoot.data.dto.ProductDto
import com.example.aroungHubKotlinSpringBoot.aroundHubKotlinSpringBoot.service.ProductService
import jakarta.validation.Valid
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/product-api")
class ProductController(
    val productService: ProductService,
    val LOGGER: Logger = LoggerFactory.getLogger(ProductController::class.java)
) {

    @GetMapping("/product/{productId}")
    fun getProduct(@PathVariable productId: String) : ProductDto{

        val startTime : Long = System.currentTimeMillis()

        LOGGER.info("[ProductController] perform {} of Aroung Hub API.", "getProduct")

        val productDto = productService.getProduct(productId)

        LOGGER.info(
            "[ProductController] Response :: productId = {}, productName = {}, productPrice = {}, productStock = {}, Response Time = {}ms",
            productDto.productId!!,
            productDto.productName!!,
            productDto.productPrice!!,
            productDto.productStock!!,
            (System.currentTimeMillis() - startTime)
        )

        return productDto
    }

    @PostMapping("/product")
    fun createProduct(@Valid @RequestBody productDto: ProductDto) : ResponseEntity<ProductDto>{
        val productId: String = productDto.productId!!
        val productName: String = productDto.productName!!
        val productPrice: Int = productDto.productPrice!!
        val productStock: Int = productDto.productStock!!

        val response : ProductDto = productService.saveProduct(productId, productName, productPrice, productStock)

        LOGGER.info(
                "[createProduct] Response >> productId : {}, productName : {}, productPrice : {}, productStock : {}",
                response.productId,
                response.productName,
                response.productPrice,
                response.productStock)

        return ResponseEntity.status(HttpStatus.OK).body(response)
    }

    @DeleteMapping("/product/{productId}")
    fun deleteProduct(@PathVariable productId: String?): ProductDto? {
        return null
    }


}